### PR TITLE
Censor private information in log files (EXPOSUREAPP-6516)

### PR DESCRIPTION
**Pull request by LukasLechnerDev**
### Description
This PR adds censoring of the following private information for our error logging: 
- PCR QR-Code (GUID, URL) 
- PCR registration token 
- Created Event Information 
- Scanned Event Information 
- Rapid Test profile information 
- Rapid Test QR-Code (URL and JSON payload (encoded and decoded))
- Rapid test registration token 
- Personal Information in Rapid test result if exist

### How to test
- Start Error Logging
- Add PCR test, add RAT test, create Event, check in to an event, create a RAT profile
- Export Error Log to your computer
- Analyze the log and check if no private information is in the log

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RACoronaTestCensor.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        with(raCoronaTest.first()) {
+            withValidName(firstName) { firstName ->
+                newMessage = newMessage.replace(firstName, "RATest/FirstName")
+            }
+
+            withValidName(lastName) { lastName ->
+                newMessage = newMessage.replace(lastName, "RATest/LastName")
+            }
+
+            val formatter = DateTimeFormat.forPattern("yyyy-MM-dd")
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RACoronaTestCensor.kt:49]: Please move this to class level or a companion object, `checkLog` will be called for every line, creating a new  `DateTimeFormat` instance for every logline is costly :money_with_wings:

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RACoronaTestCensor.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        coronaTestRepository.coronaTests.stateIn(
+            scope = debugScope,
+            started = SharingStarted.Lazily,
+            initialValue = null
+        ).filterNotNull()
+    }
+
+    override suspend fun checkLog(entry: LogLine): LogLine? {
+
+        val raCoronaTest = coronaTestFlow.map { tests -> tests.filterIsInstance<RACoronaTest>() }.first()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RACoronaTestCensor.kt:34]: this is a set right? `raCoronaTests` to show it's plural.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RatProfileCensor.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        with(ratProfile) {
+            withValidName(firstName) { firstName ->
+                newMessage = newMessage.replace(firstName, "RAT-Profile/FirstName")
+            }
+
+            withValidName(lastName) { lastName ->
+                newMessage = newMessage.replace(lastName, "RAT-Profile/LastName")
+            }
+
+            val formatter = DateTimeFormat.forPattern("yyyy-MM-dd")
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RatProfileCensor.kt:34]: Please move this to class level or a companion object, `checkLog` will be called for every line, creating a new  `DateTimeFormat` instance for every logline is costly :money_with_wings:

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RatQrCodeCensor.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+            withValidName(firstName) { firstName ->
+                newMessage = newMessage.replace(firstName, "RATest/FirstName")
+            }
+
+            withValidName(lastName) { lastName ->
+                newMessage = newMessage.replace(lastName, "RATest/LastName")
+            }
+
+            val formatter = DateTimeFormat.forPattern("yyyy-MM-dd")
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RatQrCodeCensor.kt:40]: Please move this to class level or a companion object, `checkLog` will be called for every line, creating a new  `DateTimeFormat` instance for every logline is costly :money_with_wings:

**Review by d4rken — COMMENTED** [review]
Looks very good, nice test coverage.

Few comments to improve performance.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RACoronaTestCensor.kt (hunk 1 above)
Comment (LukasLechnerDev) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RACoronaTestCensor.kt:49]: Good Point. Fixed these issues.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RACoronaTestCensor.kt (hunk 2 above)
Comment (LukasLechnerDev) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/RACoronaTestCensor.kt:36]: Yes it is. But we filter for Rapid Antigen Tests, and there should only be one, right?

I refactored it a bit.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/PcrQrCodeCensor.kt [hunk 5]
```diff
@@ -1,13 +1,14 @@
...
 import dagger.Reusable
+import de.rki.coronawarnapp.bugreporting.censors.BugCensor
 import de.rki.coronawarnapp.bugreporting.censors.BugCensor.Companion.toNewLogLineIfDifferent
 import de.rki.coronawarnapp.bugreporting.debuglog.LogLine
 import de.rki.coronawarnapp.util.CWADebug
 import javax.inject.Inject
 
 @Reusable
-class QRCodeCensor @Inject constructor() : BugCensor {
+class PcrQrCodeCensor @Inject constructor() : BugCensor {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/PcrQrCodeCensor.kt:11]: Hm, we also need to censor `CoronaTest.identifier`, as it contains the QRCode code, i.e.

```
PCRCoronaTest(identifier=qrcode-PCR-D61AC9-D61AC989-F49E-4152-880F-1AABB6716CFD, ...)
```
same thing for `RACoronaTest`.

**Review by d4rken — CHANGES_REQUESTED** [review]
Tested on device, looks good, found one more value to censor.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/PcrQrCodeCensor.kt (hunk 5 above)
Comment (LukasLechnerDev) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/bugreporting/censors/submission/PcrQrCodeCensor.kt:11]: fixed.

**Review by d4rken — DISMISSED** [review]

**Comment by jurajkusnier** [comment]
I've noticed that `TraceLocationUserInput` is not censored. When event creating fails I can see all event data in user log

```
2021-04-30T12:48:25.126Z  E/TraceLocationCreateViewModel$send: Something went wrong when sending the event TraceLocationUserInput(type=LOCATION_TYPE_PERMANENT_RETAIL, description=This Is My Description, address=This Is Address, startDate=null, endDate=null, defaultCheckInLengthInMinutes=60) to the server
java.lang.IllegalStateException: Failed to retrieve INT:CROWD_NOTIFIER_PUBLIC_KEY
	at de.rki.coronawarnapp.environment.EnvironmentSetup.getEnvironmentValue(EnvironmentSetup.kt:110)
	at de.rki.coronawarnapp.environment.EnvironmentSetup.getCrowdNotifierPublicKey(EnvironmentSetup.kt:133)
	at de.rki.coronawarnapp.presencetracing.locations.TraceLocationCreator.createTraceLocation(TraceLocationCreator.kt:20)
	at de.rki.coronawarnapp.ui.presencetracing.organizer.create.TraceLocationCreateViewModel$send$1.invokeSuspend(TraceLocationCreateViewModel.kt:73)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.lang.IllegalStateException: INT:CROWD_NOTIFIER_PUBLIC_KEY is missing in your *_environment.json
	at de.rki.coronawarnapp.environment.EnvironmentSetup.getEnvironmentValue(EnvironmentSetup.kt:107)
	... 9 more
```

**Review by jurajkusnier — APPROVED** [review]
LGTM

**Review by d4rken — APPROVED** [review]

**Pull request merged**